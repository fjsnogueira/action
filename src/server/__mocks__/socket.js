const socket = {
  emit: jest.fn(),
  setAuthToken: jest.fn()
};

export default socket;
