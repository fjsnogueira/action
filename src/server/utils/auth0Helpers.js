import {AuthenticationClient, ManagementClient} from 'auth0';
import {auth0 as auth0ClientOptions} from '../../universal/utils/clientOptions';
import getDotenv from '../../universal/utils/dotenv';

// Import .env and expand variables:
getDotenv();

export const auth0AuthenticationClient = new AuthenticationClient({
  domain: process.env.AUTH0_DOMAIN || 'parabol.auth0.com',
  // eslint-disable-next-line max-len
  token: process.env.AUTH0_MANAGEMENT_TOKEN || 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJhdWQiOiJrRGZOTURJUXRxQ21lbVFGVHVoTjh6WVN4TlNFN3JjaSIsInNjb3BlcyI6eyJ1c2VycyI6eyJhY3Rpb25zIjpbInVwZGF0ZSJdfSwidXNlcnNfYXBwX21ldGFkYXRhIjp7ImFjdGlvbnMiOlsicmVhZCIsInVwZGF0ZSIsImRlbGV0ZSIsImNyZWF0ZSJdfX0sImlhdCI6MTQ3MTkxODAzNCwianRpIjoiMzdlOWFmYjRlMjEzOWYzMzNkMDkyNjY0NDBhMzk5MDEifQ.a6NANNe5IOzvFEomZm9OeeNpRfPjoMCUzA12PaK3zVg'
});

export const auth0ManagementClient = new ManagementClient({
  domain: process.env.AUTH0_DOMAIN || 'parabol.auth0.com',
  // eslint-disable-next-line max-len
  token: process.env.AUTH0_MANAGEMENT_TOKEN || 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJhdWQiOiJrRGZOTURJUXRxQ21lbVFGVHVoTjh6WVN4TlNFN3JjaSIsInNjb3BlcyI6eyJ1c2VycyI6eyJhY3Rpb25zIjpbInVwZGF0ZSJdfSwidXNlcnNfYXBwX21ldGFkYXRhIjp7ImFjdGlvbnMiOlsicmVhZCIsInVwZGF0ZSIsImRlbGV0ZSIsImNyZWF0ZSJdfX0sImlhdCI6MTQ3MTkxODAzNCwianRpIjoiMzdlOWFmYjRlMjEzOWYzMzNkMDkyNjY0NDBhMzk5MDEifQ.a6NANNe5IOzvFEomZm9OeeNpRfPjoMCUzA12PaK3zVg'
});

export const clientId = auth0ClientOptions.clientId;

export const clientSecret = process.env.AUTH0_CLIENT_SECRET ||
  'BksPeQQrRkXhDrugzQDg5Nw-IInub9RkQ-pSWohUM9s6Oii4xoGVCrK2_OcUCfYZ';
