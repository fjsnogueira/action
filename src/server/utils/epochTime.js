export const fromEpochSeconds = (int) => {
  return new Date(int * 1000);
};

export const toEpochSeconds = (maybeDate) => {
  const msFromEpoch = maybeDate instanceof Date ? maybeDate.getTime() : maybeDate;
  return Math.floor(msFromEpoch / 1000);
};
