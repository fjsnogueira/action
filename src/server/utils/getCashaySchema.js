const path = require('path');
const resolve = require('resolve'); // eslint-disable-line import/no-extraneous-dependencies

require('babel-register')({ // eslint-disable-line import/no-extraneous-dependencies
  /*
   * Setup require and ES6 import statements to resolve from our app's
   * root path, just like our webpack environment is configured to do.
   *
   * We need this here because getCashaySchema runs from cashay,
   * it's very own node and babel context.
   *
   */
  resolveModuleSource(source, filename) {
    return resolve.sync(source, {
      basedir: path.resolve(filename, '..'),
      extensions: ['.js'],
      moduleDirectory: [
        path.join(__dirname, '..', '..'),  // application root
        path.join(__dirname, '..', '..', '..', 'node_modules'),
      ]
    });
  }
});
const {transformSchema} = require('cashay');
const graphql = require('graphql').graphql;
const rootSchema = require('../graphql/rootSchema').default;

module.exports = () => transformSchema(rootSchema, graphql);
