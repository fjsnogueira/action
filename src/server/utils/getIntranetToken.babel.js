require('babel-register');
module.exports = require('./getIntranetToken')();
