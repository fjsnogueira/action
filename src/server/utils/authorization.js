import {errorObj} from './utils';
import getRethink from '../database/rethinkDriver';
import {BILLING_LEADER} from 'universal/utils/constants';

export const getUserId = (authToken) => {
  return authToken && typeof authToken === 'object' && authToken.sub;
};

export const isSuperUser = authToken => {
  const userId = getUserId(authToken);
  return userId && authToken.rol === 'su';
};

export const requireAuth = authToken => {
  const userId = getUserId(authToken);
  if (userId) return userId;
  throw errorObj({_error: 'Unauthorized. Must be logged in for this action.'});
};

/*
 * Won't return a teamMember if it's a super user
 */
export const requireSU = authToken => {
  if (!isSuperUser(authToken)) {
    throw errorObj({_error: 'Unauthorized. Must be a super user to run this query.'});
  }
};

export const requireSUOrTeamMember = (authToken, teamId) => {
  if (!isSuperUser(authToken)) {
    const teams = authToken.tms || [];
    if (!teams.includes(teamId)) {
      throw errorObj({_error: `You do not have access to team ${teamId}`});
    }
  }
};

export const requireOrgLeaderOrTeamMember = async (authToken, teamId) => {
  const r = getRethink();
  const teams = authToken.tms || [];
  if (!teams.includes(teamId)) {
    const userId = getUserId(authToken);
    const isOrgLeader = await r.table('Team').get(teamId)('orgId').default(null)
      .do((orgId) => {
        return r.table('User').get(userId)('userOrgs')
          .filter({
            id: orgId,
            role: BILLING_LEADER
          })
          .count()
          .eq(1)
          .default(false);
      });
    if (!isOrgLeader) {
      throw errorObj({_error: `Unauthorized to view details for team ${teamId} with token ${JSON.stringify(authToken)}`});
    }
  }
};

export const requireSUOrSelf = (authToken, userId) => {
  if (isSuperUser(authToken)) return undefined;
  const authTokenUserId = getUserId(authToken);
  if (authTokenUserId === userId) {
    return userId;
  }
  throw errorObj({_error: 'Unauthorized. You cannot modify another user.'});
};

export const requireSUOrSelfOrLead = async(authToken, userId, teamId) => {
  if (isSuperUser(authToken)) return undefined;
  const authTokenUserId = getUserId(authToken);
  if (authTokenUserId === userId) {
    return 'self';
  }
  const teamMemberId = `${authTokenUserId}::${teamId}`;
  const r = getRethink();
  const teamMember = await r.table('TeamMember').get(teamMemberId);
  if (teamMember && teamMember.isLead) {
    return 'lead';
  }
  throw errorObj({_error: 'Unauthorized. Only the team member or the leader can remove someone'});
};

export const requireSUOrLead = async(authToken, teamMemberId) => {
  if (isSuperUser(authToken)) return undefined;
  const r = getRethink();
  const teamMember = await r.table('TeamMember').get(teamMemberId);
  if (!teamMember || !teamMember.isLead) {
    throw errorObj({_error: 'Unauthorized. Only the team leader promote someone to lead'});
  }
  return teamMember;
};


export const requireWebsocket = (socket) => {
  if (!socket) {
    throw errorObj({_error: 'this must be called from a websocket'});
  }
};

export const requireWebsocketExchange = (exchange) => {
  if (!exchange) {
    throw errorObj({_error: 'this requires a websocket channel exchange'});
  }
};

export const getUserOrgDoc = async(userId, orgId) => {
  const r = getRethink();
  return await r.table('User').get(userId)('userOrgs')
    .filter({id: orgId})
    .nth(0)
    .default(null);
};

export const isBillingLeader = (userOrgDoc) => {
  return (userOrgDoc && userOrgDoc.role === BILLING_LEADER);
};

export const requireOrgLeader = (userOrgDoc) => {
  const legit = isBillingLeader(userOrgDoc);
  if (!legit) {
    throw errorObj({_error: 'Unauthorized. User is not a billing leader for that organization'});
  }
};

export const requireOrgLeaderOfUser = async(authToken, userId) => {
  const r = getRethink();
  const isLeaderOfUser = await r.table('User')
    .get(authToken.sub)('userOrgs')
    .filter({
      role: BILLING_LEADER
    })
    .map((userOrg) => userOrg('id'))
    .do((leaderOrgs) => {
      return {
        leaderOrgs,
        memberOrgs: r.table('User')
          .get(userId)('userOrgs')
          .map((userOrg) => userOrg('id'))
      };
    })
    .do((res) => {
      return res('leaderOrgs')
        .union(res('memberOrgs'))
        .distinct()
        .count()
        .lt(res('leaderOrgs').count().add(res('memberOrgs').count()));
    });
  if (!isLeaderOfUser) {
    throw errorObj({_error: 'Unauthorized. Only an billing leader of a user can set this'});
  }
  return true;
};

export const requireTeamIsPaid = async(teamId) => {
  const r = getRethink();
  const isPaid = await r.table('Team').get(teamId)('isPaid').default(false);
  if (!isPaid) {
    throw errorObj({_error: `The org leader has not paid for team ${teamId}. Cannot fetch documents`});
  }
  return true;
};

// VERY important, otherwise eg a user could "create" a new team with an existing teamId & force join that team
// this still isn't secure because the resolve could get called twice & make it past this point before 1 of them writes the insert
export const ensureUniqueId = async(table, id) => {
  const r = getRethink();
  const res = await r.table(table).get(id);
  if (res) {
    throw errorObj({type: 'unique id collision'});
  }
};

export const requireUserInOrg = (userOrgDoc, userId, orgId) => {
  if (!userOrgDoc) {
    throw errorObj({_error: `Unauthorized. ${userId} does not belong in org ${orgId}`});
  }
  return true;
};

export const requireNotificationOwner = async(userId, notificationId) => {
  const r = getRethink();
  const res = await r.table('Notification').get(notificationId)('userIds').contains(userId).default(null);
  if (!res) {
    throw errorObj({_error: `Notification ${notificationId} does not exist or ${userId} does not have access to it`});
  }
  return true;
};
