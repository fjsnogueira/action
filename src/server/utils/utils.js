import mime from 'mime-types';
import {APP_MAX_AVATAR_FILE_SIZE, BILLING_LEADER} from '../../universal/utils/constants';


// Stringify an object to handle multiple errors
// Wrap it in a new Error type to avoid sending it twice via the originalError field
export const errorObj = obj => new Error(JSON.stringify(obj));

export const handleSchemaErrors = (errors, genericMessage) => {
  if (Object.keys(errors).length > 0) {
    errors._error = genericMessage || 'Server validation error';
    throw errorObj(errors);
  }
};

export function firstChange(possiblyUpdatedResult) {
  if (possiblyUpdatedResult.changes.length) {
    if (possiblyUpdatedResult.changes.length > 1) {
      console.warn('firstChange() detects more than 1 change, returning 1st.');
    }
    return possiblyUpdatedResult.changes[0];
  }
  return { new_val: undefined, old_val: undefined };
}

export function getOldVal(possiblyUpdatedResult) {
  return firstChange(possiblyUpdatedResult).old_val;
}

export function getNewVal(resultWithReturnChanges) {
  return firstChange(resultWithReturnChanges).new_val;
}

export function updatedOrOriginal(possiblyUpdatedResult, original) {
  /*
   * There will only be changes to return if there were changes made to the
   * DB. Therefore, we've got to check.
   */
  const originalReturn = (typeof original !== 'undefined') ?
    original : getOldVal(possiblyUpdatedResult);

  return getNewVal(possiblyUpdatedResult) || originalReturn;
}

// todo put this in a legitify schema
export function validateAvatarUpload(contentType, contentLength) {
  if (typeof process.env.CDN_BASE_URL === 'undefined') {
    throw errorObj({_error: 'CDN_BASE_URL environment variable is not defined'});
  }
  if (!contentType || !contentType.startsWith('image/')) {
    throw errorObj({_error: 'file must be an image'});
  }
  const ext = mime.extension(contentType);
  if (!ext) {
    throw errorObj({_error: `unable to determine extension for ${contentType}`});
  }
  if (contentLength > APP_MAX_AVATAR_FILE_SIZE) {
    throw errorObj({_error: 'avatar image is too large'});
  }
  return ext;
}

export function billingLeaderFilter(id) {
  return (orgUsers) => orgUsers.contains((orgUser) => orgUser('id').eq(id).and(orgUser('role').eq(BILLING_LEADER)));
}
