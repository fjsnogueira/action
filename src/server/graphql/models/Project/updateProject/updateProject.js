import getRethink from 'server/database/rethinkDriver';
import {ProjectInput} from 'server/graphql/models/Project/projectSchema';
import {
  GraphQLNonNull,
  GraphQLBoolean,
} from 'graphql';
import {requireSUOrTeamMember, requireWebsocket} from 'server/utils/authorization';
import shortid from 'shortid';
import ms from 'ms';
import makeProjectSchema from 'universal/validation/makeProjectSchema';
import {handleSchemaErrors} from 'server/utils/utils';

const DEBOUNCE_TIME = ms('5m');

export default {
  type: GraphQLBoolean,
  description: 'Update a project with a change in content, ownership, or status',
  args: {
    updatedProject: {
      type: new GraphQLNonNull(ProjectInput),
      description: 'the updated project including the id, and at least one other field'
    }
  },
  async resolve(source, {updatedProject}, {authToken, socket}) {
    const r = getRethink();

    // AUTH
    requireWebsocket(socket);
    // projectId is of format 'teamId::taskId'
    const [teamId] = updatedProject.id.split('::');
    requireSUOrTeamMember(authToken, teamId);

    // VALIDATION
    const schema = makeProjectSchema();
    const {errors, data: validUpdatedProject} = schema(updatedProject);
    handleSchemaErrors(errors);

    // RESOLUTION
    const {id: projectId, sortOrder, agendaId, isArchived, ...historicalProject} = validUpdatedProject;

    const now = new Date();

    const newProject = {
      ...historicalProject,
      agendaId,
      isArchived,
      sortOrder
    };
    const {teamMemberId} = historicalProject;
    if (teamMemberId) {
      const [userId] = teamMemberId.split('::');
      newProject.userId = userId;
    }
    const dbWork = [];

    if (Object.keys(updatedProject).length > 2 || sortOrder === undefined) {
      // if this is anything but a sort update, log it to history
      const mergeDoc = {
        ...historicalProject,
        updatedAt: now,
        projectId
      };
      const projectHistoryPromise = r.table('ProjectHistory')
        .between([projectId, r.minval], [projectId, r.maxval], {index: 'projectIdUpdatedAt'})
        .orderBy({index: 'projectIdUpdatedAt'})
        .nth(-1)
        .default({updatedAt: r.epochTime(0)})
        .do((lastDoc) => {
          return r.branch(
            lastDoc('updatedAt').gt(r.epochTime((now - DEBOUNCE_TIME) / 1000)),
            r.table('ProjectHistory').get(lastDoc('id')).update(mergeDoc),
            r.table('ProjectHistory').insert(lastDoc.merge(mergeDoc, {id: shortid.generate()}))
          );
        });
      dbWork.push(projectHistoryPromise);
      newProject.updatedAt = now;
    }
    dbWork.push(r.table('Project').get(projectId).update(newProject));
    await Promise.all(dbWork);
    return true;
  }
};
