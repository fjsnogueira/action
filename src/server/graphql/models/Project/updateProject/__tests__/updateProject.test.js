import getRethink from 'server/database/rethinkDriver';
import mockAuthToken from 'server/__tests__/setup/mockAuthToken';
import MockDate from 'mockdate';
import {__now, __aMinuteAgo, __anHourAgo} from 'server/__tests__/setup/mockTimes';
import fetchAndSerialize from 'server/__tests__/utils/fetchAndSerialize';
import DynamicSerializer from 'dynamic-serializer';
import MockDB from 'server/__tests__/setup/MockDB';
import expectAsyncToThrow from 'server/__tests__/utils/expectAsyncToThrow';
import socket from 'server/__mocks__/socket';
import updateProject from 'server/graphql/models/Project/updateProject/updateProject';
import {DONE} from 'universal/utils/constants';

MockDate.set(__now);
console.error = jest.fn();

describe('updateProject', () => {
  test('updates the sortOrder without changing updatedAt or making a history item', async() => {
    // SETUP
    const r = getRethink();
    const dynamicSerializer = new DynamicSerializer();
    const mockDB = new MockDB();
    const {project, user} = await mockDB.init()
      .newProject();
    const projectId = project[0].id;
    const authToken = mockAuthToken(user[7]);

    // TEST
    const updatedProject = {
      id: projectId,
      sortOrder: 2
    };
    await updateProject.resolve(undefined, {updatedProject}, {authToken, socket});

    // VERIFY
    const db = await fetchAndSerialize({
      project: r.table('Project').get(projectId),
      projectHistory: r.table('ProjectHistory')
        .between([projectId, r.minval], [projectId, r.maxval], {index: 'projectIdUpdatedAt'})
        .orderBy({index: 'projectIdUpdatedAt'})
    }, dynamicSerializer);
    expect(db).toMatchSnapshot();
  });

  test('updates the content of the project', async() => {
    // SETUP
    const r = getRethink();
    const dynamicSerializer = new DynamicSerializer();
    const mockDB = new MockDB();
    const {project, user} = await mockDB.init()
      .newProject();
    const projectId = project[0].id;
    const authToken = mockAuthToken(user[7]);

    // TEST
    const updatedProject = {
      id: projectId,
      content: 'Updated content'
    };
    await updateProject.resolve(undefined, {updatedProject}, {authToken, socket});

    // VERIFY
    const db = await fetchAndSerialize({
      project: r.table('Project').get(projectId),
      projectHistory: r.table('ProjectHistory')
        .between([projectId, r.minval], [projectId, r.maxval], {index: 'projectIdUpdatedAt'})
        .orderBy({index: 'projectIdUpdatedAt'})
    }, dynamicSerializer);
    expect(db).toMatchSnapshot();
  });

  test('updates the teamMember of the project slowly and trigger a new history item', async() => {
    // SETUP
    const r = getRethink();
    const dynamicSerializer = new DynamicSerializer();
    const mockDB = new MockDB();
    const {project, teamMember, user} = await mockDB.init()
      .newProject({updatedAt: new Date(__anHourAgo)})
      .newProjectHistory();
    const projectId = project[0].id;
    const authToken = mockAuthToken(user[7]);

    // TEST
    const updatedProject = {
      id: projectId,
      teamMemberId: teamMember[5].id
    };
    await updateProject.resolve(undefined, {updatedProject}, {authToken, socket});

    // VERIFY
    const db = await fetchAndSerialize({
      project: r.table('Project').get(projectId),
      projectHistory: r.table('ProjectHistory')
        .between([projectId, r.minval], [projectId, r.maxval], {index: 'projectIdUpdatedAt'})
        .orderBy({index: 'projectIdUpdatedAt'})
    }, dynamicSerializer);
    expect(db).toMatchSnapshot();
  });

  test('updates the content of the project quickly and do not trigger a new history item', async() => {
    // SETUP
    const r = getRethink();
    const dynamicSerializer = new DynamicSerializer();
    const mockDB = new MockDB();
    const {project, user} = await mockDB.init()
      .newProject({updatedAt: new Date(__aMinuteAgo)})
      .newProjectHistory();
    const projectId = project[0].id;
    const authToken = mockAuthToken(user[7]);

    // TEST
    const updatedProject = {
      id: projectId,
      content: 'Updated content'
    };
    await updateProject.resolve(undefined, {updatedProject}, {authToken, socket});

    // VERIFY
    const db = await fetchAndSerialize({
      project: r.table('Project').get(projectId),
      projectHistory: r.table('ProjectHistory')
        .between([projectId, r.minval], [projectId, r.maxval], {index: 'projectIdUpdatedAt'})
        .orderBy({index: 'projectIdUpdatedAt'})
    }, dynamicSerializer);
    expect(db).toMatchSnapshot();
  });

  test('updates the status of the project', async() => {
    // SETUP
    const r = getRethink();
    const dynamicSerializer = new DynamicSerializer();
    const mockDB = new MockDB();
    const {project, user} = await mockDB.init()
      .newProject();
    const projectId = project[0].id;
    const authToken = mockAuthToken(user[7]);

    // TEST
    const updatedProject = {
      id: projectId,
      status: DONE
    };
    await updateProject.resolve(undefined, {updatedProject}, {authToken, socket});

    // VERIFY
    const db = await fetchAndSerialize({
      project: r.table('Project').get(projectId),
      projectHistory: r.table('ProjectHistory')
        .between([projectId, r.minval], [projectId, r.maxval], {index: 'projectIdUpdatedAt'})
        .orderBy({index: 'projectIdUpdatedAt'})
    }, dynamicSerializer);
    expect(db).toMatchSnapshot();
  });

  test('throws when no websocket is present', async() => {
    const authToken = {};
    await expectAsyncToThrow(updateProject.resolve(undefined, {updatedProject: {}}, {authToken}));
  });

  test('throw when the caller is not a team member', async() => {
    // SETUP
    const mockDB = new MockDB();
    const {project, user} = await mockDB.init()
      .newProject();
    const authToken = mockAuthToken(user[1], {tms: ['foo']});
    const projectId = project[0].id;

    // TEST
    const updatedProject = {
      id: projectId,
      status: DONE
    };
    await expectAsyncToThrow(updateProject.resolve(undefined, {updatedProject}, {authToken, socket}), [mockDB.context.team.id]);
  });
});
