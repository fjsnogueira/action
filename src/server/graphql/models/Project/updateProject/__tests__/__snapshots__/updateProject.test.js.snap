// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`updateProject throw when the caller is not a team member 1`] = `"{\\"_error\\":\\"You do not have access to team 0\\"}"`;

exports[`updateProject throws when no websocket is present 1`] = `"{\\"_error\\":\\"this must be called from a websocket\\"}"`;

exports[`updateProject updates the content of the project 1`] = `
Object {
  "project": Object {
    "content": "Updated content",
    "createdAt": 2017-01-01T06:59:59.999Z,
    "createdBy": 1,
    "id": 0,
    "isArchived": false,
    "sortOrder": 0,
    "status": "active",
    "teamId": 2,
    "teamMemberId": 3,
    "updatedAt": 2017-01-01T08:00:00.000Z,
    "userId": 1,
  },
  "projectHistory": Array [
    Object {
      "content": "Updated content",
      "id": 4,
      "projectId": 0,
      "updatedAt": 2017-01-01T08:00:00.000Z,
    },
  ],
}
`;

exports[`updateProject updates the content of the project quickly and do not trigger a new history item 1`] = `
Object {
  "project": Object {
    "content": "Updated content",
    "createdAt": 2017-01-01T06:59:59.999Z,
    "createdBy": 1,
    "id": 0,
    "isArchived": false,
    "sortOrder": 0,
    "status": "active",
    "teamId": 2,
    "teamMemberId": 3,
    "updatedAt": 2017-01-01T08:00:00.000Z,
    "userId": 1,
  },
  "projectHistory": Array [
    Object {
      "content": "Updated content",
      "id": 4,
      "projectId": 0,
      "status": "active",
      "teamMemberId": 3,
      "updatedAt": 2017-01-01T08:00:00.000Z,
    },
  ],
}
`;

exports[`updateProject updates the sortOrder without changing updatedAt or making a history item 1`] = `
Object {
  "project": Object {
    "content": "Test Project[0]",
    "createdAt": 2017-01-01T06:59:59.999Z,
    "createdBy": 1,
    "id": 0,
    "isArchived": false,
    "sortOrder": 2,
    "status": "active",
    "teamId": 2,
    "teamMemberId": 3,
    "updatedAt": 2017-01-01T07:00:00.000Z,
    "userId": 1,
  },
  "projectHistory": Array [],
}
`;

exports[`updateProject updates the status of the project 1`] = `
Object {
  "project": Object {
    "content": "Test Project[0]",
    "createdAt": 2017-01-01T06:59:59.999Z,
    "createdBy": 1,
    "id": 0,
    "isArchived": false,
    "sortOrder": 0,
    "status": "done",
    "teamId": 2,
    "teamMemberId": 3,
    "updatedAt": 2017-01-01T08:00:00.000Z,
    "userId": 1,
  },
  "projectHistory": Array [
    Object {
      "id": 4,
      "projectId": 0,
      "status": "done",
      "updatedAt": 2017-01-01T08:00:00.000Z,
    },
  ],
}
`;

exports[`updateProject updates the teamMember of the project slowly and trigger a new history item 1`] = `
Object {
  "project": Object {
    "content": "Test Project[0]",
    "createdAt": 2017-01-01T06:59:59.999Z,
    "createdBy": 1,
    "id": 0,
    "isArchived": false,
    "sortOrder": 0,
    "status": "active",
    "teamId": 2,
    "teamMemberId": 3,
    "updatedAt": 2017-01-01T08:00:00.000Z,
    "userId": 4,
  },
  "projectHistory": Array [
    Object {
      "content": "Test Project[0]",
      "id": 5,
      "projectId": 0,
      "status": "active",
      "teamMemberId": 7,
      "updatedAt": 2017-01-01T07:00:00.000Z,
    },
    Object {
      "content": "Test Project[0]",
      "id": 6,
      "projectId": 0,
      "status": "active",
      "teamMemberId": 3,
      "updatedAt": 2017-01-01T08:00:00.000Z,
    },
  ],
}
`;
